File: packages/engine/src/facelets.ts
/**
 * Kociemba facelet strings — the interchange format for talking to anything outside this
 * engine.
 *
 * Internal piece indexing is a private convention (ours happens to match cubing.js's), and
 * other tools do not share it: GAN's smart-cube protocol reports
 * `cp = [0,5,2,1,7,4,6,3]` after `F R` where we report `[0,3,2,5,7,4,6,1]`. The 54-character
 * facelet string, by contrast, is a published standard, so it is the safe thing to exchange.
 *
 * Layout is the standard `URFDLB` order, nine facelets per face, each face read left-to-right
 * and top-to-bottom as seen looking directly at it:
 *
 * ```
 *   U: 0..8    R: 9..17   F: 18..26   D: 27..35   L: 36..44   B: 45..53
 * ```
 *
 * A solved cube in standard orientation is
 * `"UUUUUUUUURRRRRRRRRFFFFFFFFFDDDDDDDDDLLLLLLLLLBBBBBBBBB"`.
 */
import {
  CubeState,
  Face,
  NUM_CENTERS,
  NUM_CORNERS,
  NUM_EDGES,
} from "./state.ts";

/** Face letters in facelet-string order. */
const FACE_LETTERS = "URFDLB";

export const NUM_FACELETS = 54;

/**
 * Our {@link Face} indices are `U L F R B D`; the string's are `U R F D L B`. These convert
 * between the two orderings.
 */
const FACE_TO_STRING_INDEX = [0, 4, 2, 1, 5, 3] as const;
const STRING_INDEX_TO_FACE = [0, 3, 2, 5, 1, 4] as const;

/**
 * Facelets of each corner slot, in orientation order: the U/D facelet first, then the other
 * two going clockwise as seen from outside the corner. Indexed by our corner numbering
 * (URF, UBR, ULB, UFL, DFR, DLF, DBL, DRB).
 */
const CORNER_FACELETS: readonly (readonly [number, number, number])[] = [
  [8, 9, 20], // URF
  [2, 45, 11], // UBR
  [0, 36, 47], // ULB
  [6, 18, 38], // UFL
  [29, 26, 15], // DFR
  [27, 44, 24], // DLF
  [33, 53, 42], // DBL
  [35, 17, 51], // DRB
];

/**
 * Facelets of each edge slot, in orientation order. Indexed by our edge numbering
 * (UF, UR, UB, UL, DF, DR, DB, DL, FR, FL, BR, BL).
 */
const EDGE_FACELETS: readonly (readonly [number, number])[] = [
  [7, 19], // UF
  [5, 10], // UR
  [1, 46], // UB
  [3, 37], // UL
  [28, 25], // DF
  [32, 16], // DR
  [34, 52], // DB
  [30, 43], // DL
  [23, 12], // FR
  [21, 41], // FL
  [48, 14], // BR
  [50, 39], // BL
];

/** Centre facelet of each face slot, indexed by {@link Face}. */
const CENTER_FACELETS: readonly number[] = [4, 40, 22, 13, 49, 31];

export class FaceletError extends Error {
  override readonly name = "FaceletError";
}

/** The face a facelet belongs to on a solved cube — i.e. the colour of that sticker. */
const homeFaceOf = (facelet: number): number =>
  STRING_INDEX_TO_FACE[Math.floor(facelet / 9)]!;

/**
 * Render a state as a facelet string.
 *
 * Colours follow the pieces, so a rotated cube produces a rotated string: after `y`, the F
 * face shows what was the R colour. That is what a camera would see, and it is what
 * external tools expect.
 */
export function toFacelets(state: CubeState): string {
  const out = new Array<string>(NUM_FACELETS);

  for (let slot = 0; slot < NUM_CENTERS; slot++) {
    const colour = state.centers[slot]!;
    out[CENTER_FACELETS[slot]!] = FACE_LETTERS[FACE_TO_STRING_INDEX[colour]!]!;
  }

  for (let slot = 0; slot < NUM_CORNERS; slot++) {
    const piece = state.cp[slot]!;
    const orientation = state.co[slot]!;
    const here = CORNER_FACELETS[slot]!;
    const home = CORNER_FACELETS[piece]!;
    for (let i = 0; i < 3; i++) {
      // The sticker that sits at `home[i]` when solved lands on the facelet `orientation`
      // steps around this corner.
      const target = here[(i + orientation) % 3]!;
      out[target] = FACE_LETTERS[FACE_TO_STRING_INDEX[homeFaceOf(home[i]!)]!]!;
    }
  }

  for (let slot = 0; slot < NUM_EDGES; slot++) {
    const piece = state.ep[slot]!;
    const orientation = state.eo[slot]!;
    const here = EDGE_FACELETS[slot]!;
    const home = EDGE_FACELETS[piece]!;
    for (let i = 0; i < 2; i++) {
      const target = here[(i + orientation) % 2]!;
      out[target] = FACE_LETTERS[FACE_TO_STRING_INDEX[homeFaceOf(home[i]!)]!]!;
    }
  }

  return out.join("");
}

/** Corner and edge slots described by the set of faces they touch, for lookup during parsing. */
const cornerKey = (faces: readonly number[]) => [...faces].sort().join(",");

const CORNER_BY_FACES = new Map<string, number>(
  CORNER_FACELETS.map((facelets, piece) => [
    cornerKey(facelets.map(homeFaceOf)),
    piece,
  ]),
);
const EDGE_BY_FACES = new Map<string, number>(
  EDGE_FACELETS.map((facelets, piece) => [
    cornerKey(facelets.map(homeFaceOf)),
    piece,
  ]),
);

/**
 * Parse a facelet string back into a state.
 *
 * Centres are read first to establish which colour belongs to which face, so a rotated cube
 * parses correctly rather than being mistaken for a scrambled one.
 *
 * @throws {FaceletError} if the string is the wrong length, uses unknown letters, or does
 * not describe a physically coherent cube. It does *not* check solvability — a cube can be
 * coherent but unreachable (a single twisted corner), and a smart cube reporting such a
 * state is a real situation worth surfacing rather than rejecting here.
 */
export function fromFacelets(facelets: string): CubeState {
  if (facelets.length !== NUM_FACELETS) {
    throw new FaceletError(
      `expected ${NUM_FACELETS} facelets, got ${facelets.length}`,
    );
  }

  const colourAt = (facelet: number): number => {
    const index = FACE_LETTERS.indexOf(facelets[facelet]!);
    if (index === -1) {
      throw new FaceletError(
        `unknown facelet character ${JSON.stringify(facelets[facelet])} at ${facelet}`,
      );
    }
    return STRING_INDEX_TO_FACE[index]!;
  };

  const state = new CubeState();

  for (let slot = 0; slot < NUM_CENTERS; slot++) {
    state.centers[slot] = colourAt(CENTER_FACELETS[slot]!);
  }

  for (let slot = 0; slot < NUM_CORNERS; slot++) {
    const here = CORNER_FACELETS[slot]!;
    const colours = here.map(colourAt);
    const piece = CORNER_BY_FACES.get(cornerKey(colours));
    if (piece === undefined) {
      throw new FaceletError(
        `corner slot ${slot} has no such corner: ${colours.join(",")}`,
      );
    }
    // Orientation is how far the U/D-coloured sticker sits from this slot's first facelet.
    const home = CORNER_FACELETS[piece]!;
    const reference = homeFaceOf(home[0]!);
    const orientation = colours.indexOf(reference);
    if (orientation === -1) {
      throw new FaceletError(`corner slot ${slot} is missing its U/D facelet`);
    }
    state.cp[slot] = piece;
    state.co[slot] = orientation;
  }

  for (let slot = 0; slot < NUM_EDGES; slot++) {
    const here = EDGE_FACELETS[slot]!;
    const colours = here.map(colourAt);
    const piece = EDGE_BY_FACES.get(cornerKey(colours));
    if (piece === undefined) {
      throw new FaceletError(
        `edge slot ${slot} has no such edge: ${colours.join(",")}`,
      );
    }
    const home = EDGE_FACELETS[piece]!;
    state.ep[slot] = piece;
    state.eo[slot] = colours[0] === homeFaceOf(home[0]!) ? 0 : 1;
  }

  return state;
}

/** Whether two states show the same thing, comparing as facelets. */
export function faceletsEqual(a: CubeState, b: CubeState): boolean {
  return toFacelets(a) === toFacelets(b);
}

/**
 * The nine facelets of one face, reading left-to-right and top-to-bottom.
 *
 * Useful for questions about what a face *looks like* rather than where its pieces are —
 * "is this face a single colour?" is the definition of a finished OLL, and asking it this
 * way avoids the orientation conventions that make `co`/`eo` awkward for any face other
 * than U and D.
 */
export function faceletsOfFace(facelets: string, face: Face): string {
  const start = FACE_TO_STRING_INDEX[face]! * 9;
  return facelets.slice(start, start + 9);
}

/** Whether a face shows a single colour. */
export function isFaceUniform(state: CubeState, face: Face): boolean {
  const nine = faceletsOfFace(toFacelets(state), face);
  return nine === nine[0]!.repeat(9);
}

export { Face };

File: swift/CubingCore/Sources/CubingCore/Facelets.swift
/// Converting between piece arrays and the 54-character facelet string.
///
/// Facelets are the project's interchange format — the README calls it out as deliberate, because
/// it is what every other tool in cubing speaks and it does not commit anyone to GAN's or
/// cubing.js's piece indexing. Everything crossing a boundary is a facelet string: the recorded
/// vectors, the smart cube's state reports, the planner's input.
///
/// Ported from `packages/engine/src/facelets.ts`.

public enum Facelets {
    /// Face letters in facelet-string order.
    static let letters = Array("URFDLB")
    public static let count = 54

    /// Our `Face` order is `U L F R B D`; the string's is `U R F D L B`. Two orderings, one of the
    /// easier things to get quietly wrong in a port.
    static let faceToStringIndex = [0, 4, 2, 1, 5, 3]
    static let stringIndexToFace = [0, 3, 2, 5, 1, 4]

    /// Facelets of each corner slot, U/D facelet first then clockwise seen from outside.
    /// Indexed by our corner numbering: URF, UBR, ULB, UFL, DFR, DLF, DBL, DRB.
    static let cornerFacelets: [[Int]] = [
        [8, 9, 20], [2, 45, 11], [0, 36, 47], [6, 18, 38],
        [29, 26, 15], [27, 44, 24], [33, 53, 42], [35, 17, 51],
    ]

    /// Edge slots: UF, UR, UB, UL, DF, DR, DB, DL, FR, FL, BR, BL.
    static let edgeFacelets: [[Int]] = [
        [7, 19], [5, 10], [1, 46], [3, 37],
        [28, 25], [32, 16], [34, 52], [30, 43],
        [23, 12], [21, 41], [48, 14], [50, 39],
    ]

    static let centerFacelets = [4, 40, 22, 13, 49, 31]

    /// The face a facelet belongs to on a solved cube — that is, the colour of that sticker.
    static func homeFace(of facelet: Int) -> Int { stringIndexToFace[facelet / 9] }

    private static func letter(forColour colour: Int) -> Character {
        letters[faceToStringIndex[colour]]
    }

    public static func string(from state: CubeState) -> String {
        var out = [Character](repeating: " ", count: count)

        for slot in 0..<numCenters {
            out[centerFacelets[slot]] = letter(forColour: Int(state.centers[slot]))
        }

        for slot in 0..<numCorners {
            let piece = Int(state.cp[slot])
            let orientation = Int(state.co[slot])
            let here = cornerFacelets[slot]
            let home = cornerFacelets[piece]
            for i in 0..<3 {
                // The sticker that sits at `home[i]` when solved lands `orientation` steps around.
                out[here[(i + orientation) % 3]] = letter(forColour: homeFace(of: home[i]))
            }
        }

        for slot in 0..<numEdges {
            let piece = Int(state.ep[slot])
            let orientation = Int(state.eo[slot])
            let here = edgeFacelets[slot]
            let home = edgeFacelets[piece]
            for i in 0..<2 {
                out[here[(i + orientation) % 2]] = letter(forColour: homeFace(of: home[i]))
            }
        }

        return String(out)
    }

    public struct FaceletError: Error, CustomStringConvertible {
        public let description: String
    }

    private static func key(_ faces: [Int]) -> [Int] { faces.sorted() }

    private static let cornerByFaces: [[Int]: Int] = {
        var map: [[Int]: Int] = [:]
        for (piece, facelets) in cornerFacelets.enumerated() {
            map[key(facelets.map(homeFace))] = piece
        }
        return map
    }()

    private static let edgeByFaces: [[Int]: Int] = {
        var map: [[Int]: Int] = [:]
        for (piece, facelets) in edgeFacelets.enumerated() {
            map[key(facelets.map(homeFace))] = piece
        }
        return map
    }()

    /// Parse a facelet string back into a state.
    ///
    /// Centres are read first, so a rotated cube parses as rotated rather than as scrambled. Like
    /// the TypeScript this checks coherence, not solvability: a twisted corner is a real thing a
    /// smart cube can report, and worth surfacing rather than rejecting here.
    public static func state(from facelets: String) throws -> CubeState {
        let chars = Array(facelets)
        guard chars.count == count else {
            throw FaceletError(description: "expected \(count) facelets, got \(chars.count)")
        }
        func colour(_ at: Int) throws -> Int {
            guard let index = letters.firstIndex(of: chars[at]) else {
                throw FaceletError(description: "unknown facelet character \(chars[at]) at \(at)")
            }
            return stringIndexToFace[index]
        }

        var state = CubeState.solved
        for slot in 0..<numCenters { state.centers[slot] = UInt8(try colour(centerFacelets[slot])) }

        for slot in 0..<numCorners {
            let colours = try cornerFacelets[slot].map(colour)
            guard let piece = cornerByFaces[key(colours)] else {
                throw FaceletError(description: "corner slot \(slot) has no such corner")
            }
            let reference = homeFace(of: cornerFacelets[piece][0])
            guard let orientation = colours.firstIndex(of: reference) else {
                throw FaceletError(description: "corner slot \(slot) is missing its U/D facelet")
            }
            state.cp[slot] = UInt8(piece)
            state.co[slot] = UInt8(orientation)
        }

        for slot in 0..<numEdges {
            let colours = try edgeFacelets[slot].map(colour)
            guard let piece = edgeByFaces[key(colours)] else {
                throw FaceletError(description: "edge slot \(slot) has no such edge")
            }
            state.ep[slot] = UInt8(piece)
            state.eo[slot] = colours[0] == homeFace(of: edgeFacelets[piece][0]) ? 0 : 1
        }
        return state
    }
}
